src/System.Buffers/src/System/Buffers/BufferBucket.cs

```diff
@@ -0,0 +1,76 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Diagnostics;
+using System.Threading;
+
+namespace System.Buffers
+{
+    internal sealed class BufferBucket
+    {
+        private volatile int _index;
+        private byte[][] _data;
+        private int _bufferSize;
+        private SpinLock _lock;
+
+        internal BufferBucket(int bufferSize, int numberOfBuffers)
+        {
+            _index = 0;
+            _bufferSize = bufferSize;
+            _lock = new SpinLock();
+            _data = new byte[numberOfBuffers][];
+            for (int i = 0; i < numberOfBuffers; i++)
+            {
+                _data[i] = new byte[bufferSize];
```

Comment:
I went back and forth on this one and went with the upfront allocation so that all calls to Rent would (assuming we had free buffers of the requested size) be constant-time calls; if we had the buffers be lazy allocated, you would lose efficiency since you would have to go and allocate the buffers on first response. My thought was that it would be better to push the cost up-front and have constant-time rent calls, especially since the Pool itself is an instance-object...so they wouldn't create a pool with certain buffer sizes unless the caller knew they needed one.